### PR TITLE
bug: unable to list newly purchased kittyitem

### PR DESCRIPTION
**Pull request by MrDSGC**
- fixed issue where ListingAvailable event was fetching a listing that was already completed
- added some documentation for api classes
- dependency updates

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -8,6 +8,7 @@
     "chalk-animation": "^1.6.0",
     "cross-env": "^7.0.3",
     "dotenv": "^16.0.0",
+    "flow": "^0.2.3",
```
Comment (psiemens) [inline package.json:11]: I don't see this dependency being used anywhere. Did you mean to add it?

File: api/src/workers/listing-handler.ts [hunk 2]
```diff
@@ -1,9 +1,14 @@
+/**
+ * Listing Event Handler
+ *
+ * This worker extends BaseEventHandlr & listens for Listing events from flow.
```
Comment (psiemens) [inline api/src/workers/listing-handler.ts:4]:
```suggestion
 * This worker extends BaseEventHandler & listens for Listing events from Flow.
```

File: api/src/services/storefront.ts [hunk 3]
```diff
@@ -1,3 +1,10 @@
+/**
+ * Storefront Service
+ * 
+ * Service class used to execute scripts & transactions on the flow blockchain to buy & sell kitty items.
```
Comment (psiemens) [inline api/src/services/storefront.ts:4]:
```suggestion
 * Service class used to execute scripts & transactions on the Flow blockchain to buy & sell Kitty Items. 
```

File: api/src/routes/storefront.ts [hunk 4]
```diff
@@ -1,3 +1,9 @@
+/**
+ * Storefront Router
+ * 
+ * Router class that defines api REST endpoints and calls storefront service with request data.
```
Comment (psiemens) [inline api/src/routes/storefront.ts:4]:
```suggestion
 * Router class that defines API REST endpoints and calls storefront service with request data.
```

**Review by psiemens — COMMENTED** [review]
Looks good! 

Thanks so much for adding and improvement the doc comments 🚀

Just have one question about the `flow` JS dependency.

File: package.json (hunk 1 above)
Comment (MrDSGC) [reply package.json]: removed

**Review by psiemens — APPROVED** [review]
LGTM!

File: api/src/services/storefront.ts [hunk 5]
```diff
@@ -172,28 +179,28 @@ class StorefrontService {
...
+            price: item.price,
+            transaction_id: listingEvent.transactionId,
+          })
+          .returning("transaction_id")
+          .onConflict("listing_resource_id")
+          .ignore()
+          .catch(e => {
+            console.log(e)
+          })
+      })
```
Comment (psiemens) [inline api/src/services/storefront.ts:202]: Looks like there's an extra level of indentation here

File: api/src/workers/base-event-handler.ts [hunk 6]
```diff
@@ -16,101 +21,80 @@ abstract class BaseEventHandler {
...
+      // currentBlockHeight in the DB will equal the toBlock of the previous blockrange fcl query.
+      // increment fromBlock by 1 for next window.
+      let fromBlock = blockCursor.currentBlockHeight + 1;
+      let toBlock = await this.flowService.getLatestBlockHeight();
+
+      // on 1st iteration, fromBlock will be greater than toBlock due to newly inserted row.
+      if (fromBlock > toBlock) {
+        fromBlock = toBlock;
+      }
+      console.log(`Checking Block range fromBlock=${fromBlock} toBlock=${toBlock}`);
```
Comment (psiemens) [inline api/src/workers/base-event-handler.ts:45]:
```suggestion
      console.log(`Checking block range: fromBlock=${fromBlock} toBlock=${toBlock}`);
```

File: api/src/workers/base-event-handler.ts [hunk 7]
```diff
@@ -16,101 +21,80 @@ abstract class BaseEventHandler {
...
+      if (events.length > 0) {
+        // make sure events are sorted by blockHeight ascending
+        events.sort((event1, event2) => {
+          if (event1.blockHeight > event2.blockHeight) {
+            return 1;
+          } else if (event1.blockHeight < event2.blockHeight) {
+            return  -1;
+          } 
+          return 0;
+        });
```
Comment (psiemens) [inline api/src/workers/base-event-handler.ts:75]:
```suggestion
        events.sort((event1, event2) => event1.blockHeight - event2.blockHeight);
```

I think this should be equivalent

Comment (psiemens) [reply api/src/workers/base-event-handler.ts:75]: Does this sort function keep the events sorted by name as well?

I'm worried about a case where this sort function may shuffle events within the same block. 

For example:

```js
const events = [
  { blockHeight: 6, name: "ListingAvailable },
  { blockHeight: 10, name: "ListingAvailable" },
  { blockHeight: 8, name: "ListingCompleted" },
  { blockHeight: 10, name: "ListingCompleted" }
]

console.log(sortByBlockHeight(events))

//  [
//    { blockHeight: 6, name: "ListingAvailable },
//    { blockHeight: 8, name: "ListingCompleted" },
//    { blockHeight: 10, name: "ListingCompleted" }, <--- OUT OF ORDER, available must come before complete
//    { blockHeight: 10, name: "ListingAvailable" },
//  ]
```

**Review by psiemens — COMMENTED** [review]
Looks really good! Very clean code ✨ 

I'm still worried about another edge case (copied below from a previous message). However, I think we should verify + fix that case in a separate PR.

--- 

Example scenario:

- `ListingAvailable` event is detected in block 19.
- This calls `StorefrontService.addListing`, which in turn executes the `get_listing_item.cdc` Cadence script on the emulator. It calls this script because it only has the listing ID -- it needs more information about the listing to insert into the database.
- This uses `fcl.query`, [which by default executes the script against the state from the current latest block](https://docs.onflow.org/http-api/#tag/Scripts/paths/~1scripts/post).
- The kicker is: **by this point, the latest block is 20**, which is the block at which that listing was removed. The script returns empty because the listing no longer exists at that version of the blockchain state.

So if this is a problem (I'm just guessing at this point), one possible solution would be to make the event processors "block height aware". i.e. If a `ListingAvailable` event is detected at block 19, then the script to fetch the listing is also executed at block 19.

**Comment by MrDSGC** [comment]
> Looks really good! Very clean code ✨
> 
> I'm still worried about another edge case (copied below from a previous message). However, I think we should verify + fix that case in a separate PR.
> 
> Example scenario:
> 
> * `ListingAvailable` event is detected in block 19.
> * This calls `StorefrontService.addListing`, which in turn executes the `get_listing_item.cdc` Cadence script on the emulator. It calls this script because it only has the listing ID -- it needs more information about the listing to insert into the database.
> * This uses `fcl.query`, [which by default executes the script against the state from the current latest block](https://docs.onflow.org/http-api/#tag/Scripts/paths/~1scripts/post).
> * The kicker is: **by this point, the latest block is 20**, which is the block at which that listing was removed. The script returns empty because the listing no longer exists at that version of the blockchain state.
> 
> So if this is a problem (I'm just guessing at this point), one possible solution would be to make the event processors "block height aware". i.e. If a `ListingAvailable` event is detected at block 19, then the script to fetch the listing is also executed at block 19.

new ticket to address this created https://app.zenhub.com/workspaces/flow-developer-experience-6052fcad2959f80011fda14b/issues/onflow/kitty-items/259

**Pull request merged**